### PR TITLE
[SPARK-13339] [DOCS] Clarify commutative / associative operator requirements for reduce, fold

### PR DESCRIPTION
**Pull request by srowen**
Clarify that reduce functions need to be commutative, and fold functions do not

See https://github.com/apache/spark/pull/11091

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 1]
```diff
@@ -461,10 +461,10 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
     fromRDD(rdd.partitionBy(partitioner))
 
   /**
-   * Merge the values for each key using an associative reduce function. This will also perform
-   * the merging locally on each mapper before sending results to a reducer, similarly to a
-   * "combiner" in MapReduce.
-   */
+    * Return an RDD containing all pairs of elements with matching keys in `this` and `other`. Each
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: This is not directly related; I just found this doc copy and paste error and fixed it along the way.

**Comment by YPares** [comment]
@srowen Fold function do not need to be commutative? I thought it was the case only in Python?

**Comment by srowen** [comment]
They don't need to be since `fold` will operate on the input in some order. Spark's `fold` is built on `Iterator.fold` which would naturally apply the operator in a fixed order from first to last, so commutativity does not matter. See its doc: http://www.scala-lang.org/api/2.11.7/index.html#scala.collection.Iterator ; most Spark docs for `fold` methods do not mention commutativity either.

The scaladoc does mention the non-trivial things that happen if it's not commutative, but it's not required.

File: R/pkg/R/RDD.R [hunk 2]
```diff
@@ -632,10 +632,10 @@ setMethod("Filter",
 #' Reduce across elements of an RDD.
 #'
 #' This function reduces the elements of this RDD using the
-#' specified commutative and associative binary operator.
+#' specified commutative and associative and commutative binary operator.
```
Comment (srowen) [inline R/pkg/R/RDD.R]: Ugh, I just realized I added one too many commutatives in a few places.

**Comment by srowen** [comment]
Merged to master
